CRM/Contact/Form/Task/PDFLetterCommon.php

```diff
@@ -396,16 +410,53 @@ public static function postProcess(&$form) {
       $html[] = $tokenHtml;
     }
 
+    $saveToFile = count($activityIds) === 1 && Civi::settings()->get('recordGeneratedLetters') === 'combined-attached';
+    if ($saveToFile) {
+      // We need to capture data from various output-generators whose only
+      // consistent output format is writing to the console (for download).
+      $tmpFile = tempnam(sys_get_temp_dir(), 'PDFLetterCommon-');
+      $tmpFh = fopen($tmpFile, 'w');
+      ob_start(function ($buf) use ($tmpFh) {
+        // Pass to file, then continue to stdout.
+        fwrite($tmpFh, $buf);
+        return FALSE;
+      });
```

Comment:
> why is `ob_end_flush()` and `fclose()` called later?

In the intervening lines (roughly 430-445), the functions `CRM_Utils_PDF_Utils::html2pdf()`, `CRM_Utils_PDF_Document::printDocuments()`, and `CRM_Utils_PDF_Document::html2doc()` all print documents to the console. The goal here is to capture their output and CC to a file. (A bit like [`tee`](https://en.wikipedia.org/wiki/Tee_(command)).)

> ... a class / service responsible for writing content to a temporary file...

At first I thought it wouldn't make the code much clearer. But... this function is bordering on [toxic](https://wiki.civicrm.org/confluence/display/CRM/Toxic+Code+Protocol), and I wasn't able to do a deeper cleaning, and we should do something to hold-down the toxicity. Extracting that class feels like the least we can do.